### PR TITLE
Matrix norm

### PR DESCRIPTION
**Pull request by kswirydo**
Matrix inf norm commonly appears in error evaluation formulae (i.e. NSR, etc). I added Cuda, CPU and HIP versions. It is not a part of standard BLAS.

Matrix Inf norm is max of the sum of rows (with abs value).
This is done as 3 kernels:
1) Sum rows, store in `d_r` (buffer).
2) Reduce `d_r` to another buffer (blockwise).
3) Second pass of reduction - reduce to a variable.
It is pretty straightforward to do on CPU.

There is a `cusolverSp` function that does 2+3 (which I am using). The rest I implemented myself. 
Should be easy to add unit test (?) Can add before merging.

### CODE REVIEW

File: resolve/matrix/MatrixHandlerHip.cpp [hunk 1]
```diff
@@ -103,9 +104,39 @@ namespace ReSolve {
...
   }
 
-  int MatrixHandlerHip::Matrix1Norm(matrix::Sparse* /* A */, real_type* /* norm */)
+  int MatrixHandlerHip::MatrixInfNorm(matrix::Sparse* A, real_type* norm)
   {
-    return -1;
+    // we assume A is in CSR format
+    real_type* d_r = workspace_->getDr();
+    printf("d_r size %d \n", workspace_->getDrSize());
+    if (workspace_->getDrSize() != A->getNumRows()) {
```
Comment (pelesh) [inline resolve/matrix/MatrixHandlerHip.cpp:112]: This print statement violates xSDK requirements.

File: resolve/LinSolverDirectCuSolverRf.cpp [hunk 2]
```diff
@@ -25,11 +26,28 @@ namespace ReSolve
...
+      mem_.allocateArrayOnDevice(&d_P_, n);
+    } 
+
+    if (d_Q_ == nullptr){
+      mem_.allocateArrayOnDevice(&d_Q_, n);
+    }
+
+    if (d_T_ != nullptr){
+      mem_.deleteOnDevice(d_T_);
+    }
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/LinSolverDirectCuSolverRf.cpp:49]: After looking closer at the code, my suggestion would not work.

File: resolve/matrix/MatrixHandlerHip.cpp (hunk 1 above)
Comment (kswirydo) [reply resolve/matrix/MatrixHandlerHip.cpp:112]: It was placed here for debugging/testing only.

File: resolve/vector/VectorHandler.cpp [hunk 3]
```diff
@@ -129,6 +129,30 @@ namespace ReSolve {
     }
   }
 
+  /** 
+   * @brief compute infinity norm of a vector (i.e., find an entry with largest absolute value)
+   * 
+   * @param[in] The vector
+   * @param memspace[in] string containg memspace (cpu or cuda or hip)
+   *
```
Comment (pelesh) [inline resolve/vector/VectorHandler.cpp:137]:
```suggestion
   * @param[in] The vector
   * @param[in] memspace string containg memspace (cpu or cuda or hip)
   *
```

**Review by pelesh — APPROVED** [review]
Looks good!

**Pull request merged**